### ISSUE TITLE
[arrays] vector-matrix multiplication + type of the slice of a matrix

### ISSUE THREAD

**Issue by tayral** [TRIQS/triqs#405]
Hi, 
Working with the old "dev", i.e more or less "unstable":
It seems that vector-matrix multiplication is not implemented:
```c++
  tqa::vector<double> v = {1.,2.};
  tqa::matrix<double> M = {{1.,3.},{4.,5.}};
 
  auto prod = M*v;  //matrix*vector compiles and returns {7,14}, as it should
  auto prod2 = v*M; //vector*matrix does not compile
```
Thus, TRIQS seems to interpret a triqs::vector as a column vector. Unless I missed it, this is not documented... 
Aside from this, it seems to me that the expression prod2 is unambiguous, and should therefore compile and return {9, 13}.
I encountered this issue while implementing the following operation:
```c++
 auto prod3 = M1(range(),4).transpose()*M2; //assuming M1 and M2 are nxn matrices
```
This does not compile because M1(range(),4) is a vector of size n instead of a matrix of shape (n, 1), and hence
- it does not make sense to transpose it to make it a 1xn matrix
- because of the above problem with vector-matrix multiplication, the * operation is not defined

To sum up, there are two issues/questions:
- Should M1(range(),4) return a matrix or a vector? Were it to return a matrix (like it does in numpy), my expression prod3 would compile.
- If it returns a vector (as it now does), should we allow for vector-matrix multiplication?

Thanks
Thomas

**Comment by parcollet** [comment]
yes, it was not implemented, it would be easy to add, but indeed a vector is a column...
@mferrero : opinion ?

**Comment by parcollet** [comment]
Decision: 
matrix slice should be matrices.
IT also simplify code.
Will do it with array upgrade in 1.7

**Comment by Wentzell** [comment]
This concerns the old `triqs::arrays` library, which has since been removed and replaced by the standalone **`nda`** array library — `c++/triqs/arrays/` no longer exists, and `triqs/arrays.hpp` is now just a backward-compatibility umbrella header that includes `nda`. The `tqa::vector`/`tqa::matrix` types and operators referenced here are gone.

On the matrix-slice question: in `nda` the result rank follows the numpy convention — a scalar (integer) index drops that dimension, while a `range` preserves it. So the two cases are distinct *by syntax*:

- `M(range(), 4)` — the integer `4` reduces the rank, giving a **rank-1 vector** of length *n* (tayral's original slice).
- `M(range(), range(4,5))` — both arguments are ranges, so the rank is preserved, giving the **n×1 matrix** that can be transposed and chained.

So the "should a matrix slice be a matrix or a vector" choice is now in the user's hands via the index type. Matrix/vector product semantics are likewise defined by `nda` now.

Closing as obsolete here. If the vector×matrix product or slicing behavior is still a concern with the current array library, please open an issue against [TRIQS/nda](https://github.com/TRIQS/nda).
